crates/compiler/unify/src/unify.rs

```diff
@@ -2978,8 +2978,18 @@ fn unify_flat_type<M: MetaCollector>(
         (Apply(l_symbol, l_args), Apply(r_symbol, r_args)) if l_symbol == r_symbol => {
             let mut outcome = unify_zip_slices(env, pool, *l_args, *r_args, ctx.mode);
 
+            let mut chosen_args = Vec::with_capacity(l_args.len());
+
+            for (var1, var2) in l_args.into_iter().zip(r_args.into_iter()) {
+                let var1 = env.subs[var1];
+                let var2 = env.subs[var2];
+                chosen_args.push(choose_merged_var(env.subs, var1, var2));
+            }
+
+            let chosen_args = SubsSlice::extend_new(&mut env.subs.variables, chosen_args);
```

Comment:
Let's avoid an intermediate allocation here. You can allocate a zeroed-slice and then fill it in:

```suggestion
            let chosen_args = SubsSlice::reserve_into_subs(&mut env.subs, l_args.len());

            for ((store, var1), var2) in chosen_args.iter().zip(l_args).zip(r_args) {
                let var1 = env.subs[var1];
                let var2 = env.subs[var2];
                env.subs[store] = choose_merged_var(env.subs, var1, var2);
            }
```

Reply:
(applied, with some fixups)